src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -220,6 +220,14 @@ private void Dispose(bool disposing)
             }
         }
 
+        private void EnsureMessageSizeIsValid(Int32 messageSize)
+        {
+            if (messageSize < 0 || messageSize > _description.MaxMessageSize)
```

Comment:
If I'm understanding the code correctly, that's dealt with here: https://github.com/mongodb/mongo-csharp-driver/blob/fad2887cd28950ebb50d1039e336026d1f169339/src/MongoDB.Driver.Core/Core/Connections/IsMasterResult.cs#L215

Reply:
>If I'm understanding the code correctly, that's dealt with here

yes, but this value can be used only when we had at least one `isMaster` response.

When we call `isMaster` first time, `_description` will be `null`, since we haven't had `isMaster` response yet and as a result, we haven't filled `_description` yet. So, since `description` is `null` we will have a null reference exception in this line. 
See test results: https://evergreen.mongodb.com/version/5d10da1e9ccd4e148cd09c92 :) 

Also, as I see, the quoted test says that if you haven't had `isMaster` response yet(`e.g. the first ismaster command sent on first connection`), you need to use hardcoded value in MaxMessageSize which is 48 MB, but as soon as the `ismaster` response is received, you will need to use the value from `isMaster` response (`once you have received a valid ismaster response.`)

Reply:
> you need to use hardcoded value in MaxMessageSize which is 48 MB

I think the spec is wrong. When we haven't received an isMaster result yet we need to use a default value of 16000000 to be compatible with older servers.

In any case it's probably moot... since I can't imagine an isMaster command or response reaching even the smaller max message size.

But the NullReferenceException needs to be fixed... :)

Reply:
Dmitry argues that we should just follow the spec and use 48000000 (and not 16000000).

While it is totally true that the limit used to be 16000000, looking at server JIRA tickets it has been 48000000 since at least 2012-12-26 when the `maxMessageSizeBytes` field was first added to the `isMaster` result.

So it seems safe to ignore the fact that the `MaxMessageSize` once used to be smaller.

You can create a local variable here to work around the `NullReferenceException`:

     var maxMessageSize = _description?.MaxMessageSize ?? 48000000;

Reply:
> I think the spec is wrong. When we haven't received an isMaster result yet we need to use a default value of 16000000 to be compatible with older servers.

Just to clarify, when _description is null, I should be using a hard-coded value of 16MB as maxMessageSize?

Reply:
Go ahead and just follow the spec and use 48000000.

No servers with the smaller 16000000 MasMessageSize should still exist in the wild.

And as Dmitry argues it's easier to just follow the spec than to try and change the spec.

Reply:
p.s. 48MB and 48000000 are not the same value.

I tend to interpret 48MB to mean 48 * 1024 * 1204 = 50331648.